modules/apps/29-fee/keeper/escrow.go

```diff
@@ -44,66 +44,143 @@ func (k Keeper) EscrowPacketFee(ctx sdk.Context, identifiedFee *types.Identified
 }
 
 // DistributeFee pays the acknowledgement fee & receive fee for a given packetId while refunding the timeout fee to the refund account associated with the Fee
-func (k Keeper) DistributeFee(ctx sdk.Context, refundAcc, forwardRelayer, reverseRelayer sdk.AccAddress, packetID *channeltypes.PacketId) error {
+func (k Keeper) DistributeFee(ctx sdk.Context, refundAcc, forwardRelayer, reverseRelayer string, packetID *channeltypes.PacketId) error {
+	var packetIdErr error
+	var distributeFwdErr error
+	var distributeReverseErr error
+	var refundAccErr error
+
 	// check if there is a Fee in escrow for the given packetId
 	feeInEscrow, found := k.GetFeeInEscrow(ctx, packetID)
 	if !found {
-		return sdkerrors.Wrapf(types.ErrFeeNotFound, "with channelID %s, sequence %d", packetID.ChannelId, packetID.Sequence)
+		packetIdErr = sdkerrors.Wrapf(types.ErrFeeNotFound, "with channelID %s, sequence %d", packetID.ChannelId, packetID.Sequence)
 	}
 
-	// send receive fee to forward relayer
-	err := k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, forwardRelayer, feeInEscrow.Fee.ReceiveFee)
-	if err != nil {
-		return sdkerrors.Wrap(err, "failed to send fee to forward relayer")
+	// cache context before trying to send to forward relayer
+	cacheCtx, writeFn := ctx.CacheContext()
+
+	fwd, err := sdk.AccAddressFromBech32(forwardRelayer)
+	if err == nil {
+		err = k.bankKeeper.SendCoinsFromModuleToAccount(cacheCtx, types.ModuleName, fwd, feeInEscrow.Fee.ReceiveFee)
+		if err == nil {
+			// write the cache
+			writeFn()
+			// NOTE: The context returned by CacheContext() refers to a new EventManager, so it needs to explicitly set events to the original context.
+			ctx.EventManager().EmitEvents(cacheCtx.EventManager().Events())
+		}
+	} else {
+		distributeFwdErr = sdkerrors.Wrap(err, "failed to send fee to forward relayer")
 	}
 
-	// send ack fee to reverse relayer
-	err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, reverseRelayer, feeInEscrow.Fee.AckFee)
-	if err != nil {
-		return sdkerrors.Wrap(err, "error sending fee to reverse relayer")
+	// cache context before trying to send to reverse relayer
+	cacheCtx, writeFn = ctx.CacheContext()
+
+	reverse, err := sdk.AccAddressFromBech32(reverseRelayer)
+	if err == nil {
+		err = k.bankKeeper.SendCoinsFromModuleToAccount(cacheCtx, types.ModuleName, reverse, feeInEscrow.Fee.AckFee)
+		if err == nil {
+			// write the cache
+			writeFn()
+			// NOTE: The context returned by CacheContext() refers to a new EventManager, so it needs to explicitly set events to the original context.
+			ctx.EventManager().EmitEvents(cacheCtx.EventManager().Events())
+		}
```

Comment:
This is duplicated 3 times. The only difference is the arguments provided to `SendCoinsFromModuleToAccount`. Lets abstract into a function:

```go 
func (k Keeper) DistributeFee(ctx sdk.Context, receiver sdk.AccAddress, fee sdk.Coins) {
        // cache context before trying to send to reverse relayer
	cacheCtx, writeFn = ctx.CacheContext()
	
	err = k.bankKeeper.SendCoinsFromModuleToAccount(cacheCtx, types.ModuleName, receiver, fee)
	if err == nil {
		// write the cache
		writeFn()
		
		// NOTE: The context returned by CacheContext() refers to a new EventManager, so it needs to explicitly set events to the original context.
		ctx.EventManager().EmitEvents(cacheCtx.EventManager().Events())
	}
}
```

```go
func (k Keeper) DistributePacketFees(ctx sdk.Context, refundAcc, forwardRelayer string, reverseRelayer sdk.AccAddress, feeInEscrow types.IdentifiedPacketFee) {
    // distribute fee for forward relaying 
    forward, err := sdk.AccAddressFromBech32(forwardRelayer) // optionally emit event for err or log error
    if err == nil {
        k.DistributeFee(ctx, forwardRelayer, feeInEscrow.Fee.ReceiveFee)
    }
    
    // distribute fee for reverse relaying
    k.DistributeFee(ctx, reverse, feeInEscrow.Fee.AckFee)
    
    // refund timeout fee
    refund, err := sdk.AccAddressFromBech32(refundAcc) // optionally panic on error here since refundAcc should always be parseable
    if err == nil {
        k.DistributeFee(ctx, refundAcc, feeInEscrow.Fee.TimeoutFee)
    }
}
```